### PR TITLE
Fix rooms being in stripped and non-stripped stores

### PR DESCRIPTION
**Pull request by FlixCoder**
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/978 , potentially related to #505 ?

I would like to ask whether this could break something?

Also there might be additional changes needed for stripped states vs. states in the stores?

I added some cleaning up that I had laying around as well, I hope this is fine.

### CODE REVIEW

**Comment by FlixCoder** [comment]
So I actually added the state as well to the MR, because it was needed to fully fix the bug

**Comment by FlixCoder** [comment]
So ideally, the sled and indexeddb stores would also delete the whole (stripped) room data, if the respective other (non-stripped/stripped) was added, but that is difficult to do. The transaction trees don't allow to search for keys, so deleting the room data would need to happen outside of the transaction, which is not optimal. It would also need another iteration of running through the changes potentially. I am not yet sure how to solve or whether it is necessary, as it already seems to fix my bug at least this way.

**Comment by zecakeh** [comment]
I'm actually wondering if it makes sense to still have separated room lists, since they now both use the same `RoomInfo` type. It would avoid that kind of issue where a room is duplicated between the lists.

We could have a single accessor, maybe with a bitset to filter rooms, similar to what was proposed in #930.

**Comment by chrisguida** [comment]
FWIW I tested this with the getting_started example on [this PR](https://github.com/matrix-org/matrix-rust-sdk/pull/977) and now I can kick/reinvite the bot as many times as I want and it successfully autojoins!

**Comment by FlixCoder** [comment]
> I'm actually wondering if it makes sense to still have separated room lists, since they now both use the same `RoomInfo` type. It would avoid that kind of issue where a room is duplicated between the lists.
> 
> We could have a single accessor, maybe with a bitset to filter rooms, similar to what was proposed in #930.

Might make sense yeah, but would be a more complicated task and would break things right? Would suggest to do that in another MR, as changing all methods to only check the one room infos sounds like a different thing.

**Comment by zecakeh** [comment]
Yes, that would also mean we have to provide migrations for the databases.

**Comment by FlixCoder** [comment]
> So ideally, the sled and indexeddb stores would also delete the whole (stripped) room data, if the respective other (non-stripped/stripped) was added, but that is difficult to do. The transaction trees don't allow to search for keys, so deleting the room data would need to happen outside of the transaction, which is not optimal. It would also need another iteration of running through the changes potentially. I am not yet sure how to solve or whether it is necessary, as it already seems to fix my bug at least this way.
> 
> EDIT: I have a solution for the sled store ready, just unsure how good it is. Will wait for feedback before trying the same on indexedDB to see whether that is something that is sensible. (Visible in https://github.com/FlixCoder/matrix-rust-sdk/tree/fixes)

I would like to ask regarding this again: what do you think about removing the whole room data and the way it is done? Which way do you prefer, the PR here or the other branch?

I fixed the last things now too, so the PR would be ready from my perspective.

**Comment by gnunicorn** [comment]
thanks for the PR @FlixCoder, especially adding tests and integration tests!

However, removing the data from the databases isn't really fixing the actual issue, just mitigating it because there is a fallback on the bucket to pick the state from. Storing and retrieving the data correctly for all states isn't particularly easy to do ([see the section in the spec](https://spec.matrix.org/v1.3/client-server-api/#stripped-state)). In particular when leaving or being kicked from a room, we do want to keep the state around to make it possible for the user to still understand what the room was about. Coming to think of it, there is no clear scenario where we _must_ remove either state from the stores, we merely have to pick the correct ones.

Your tests do show though that we have a bug _picking_ the correct bucket to retrieve the data from. The solution is not to remove the data, though, but instead we need to figure out the scenarios and reasons why we are picking the wrong bucket and make sure to always pick the right bucket.


Sorry, for the lengthy response, it is not easy to explain this briefly. I hope you understand the problem and why we don't want to blanket remove the data. Are you interested in using the tests to find the actual culprit?

**Comment by FlixCoder** [comment]
How would the correct one be picked though? If both are present, do they always have timestamps to compare? Should all stores act similarly? As they current do not act similarly in all cases and use different data.

Also should stripped state be removed when adding proper state, as that is said in the spec as you linked?

**Comment by FlixCoder** [comment]
So @gnunicorn I assumed that given the spec it is ok to remove stripped data once we get regular data. Based on that it was possible to easily change the retrieving logic as the decision is as simple as it was before: use stripped state over non-stripped. I think it might be quite complicated to keep both in all scenarios, as the decision must be based on events that need to be deserialized first and such.

The PR is updated and should work this way for you hopefully :)

File: testing/matrix-sdk-integration-testing/assets/ci-start.sh [hunk 1]
```diff
@@ -8,6 +8,8 @@ echo " ====== Patching for CI  ====== "
 sed -i 's/^#enable_registration_without_verification:.*$/enable_registration_without_verification: true/g' /data/homeserver.yaml
 sed -i 's/^#enable_registration:.*$/enable_registration: true/g' /data/homeserver.yaml
 echo """
+enable_registration: true
+enable_registration_without_verification: true
```
Comment (gnunicorn) [inline testing/matrix-sdk-integration-testing/assets/ci-start.sh:13]: if we add them, we don't have to `sed` ...
```suggestion
echo """
enable_registration: true
enable_registration_without_verification: true

```

**Pull request merged**